### ISSUE TITLE
Ensure backward() is idempotent

### ISSUE THREAD

**Issue by eggsyntax** [karpathy/micrograd#53]
Hi Andrej,

Many thanks for micrograd & its accompanying video; they deepened my understanding of backprop considerably!

I notice that in the current implementation, calling `backward()` repeatedly is non-idempotent, because the grads just keep accumulating. This seems like something people are likely to trip over. The fix is simple: in the def of `backward()`, just above
```python
        # go one variable at a time and apply the chain rule to get its gradient
        self.grad = 1
```
add
```python
        # reset gradients to ensure they don't get repeatedly accumulated
        for v in reversed(topo):
            v.grad = 0
```

Just submitted [PR 54](https://github.com/karpathy/micrograd/pull/54) for your consideration which just makes that one change.

Example of non-idempotence with current master branch: given a simple tree where a = 3, b = 2, c = a + b, d = 1, e = c * d (all leaves as Values of course):

```python
>>> print_grads()
print_grads()
a: 0, b: 0, c: 0, d: 0, e: 0
>>> e.backward()
e.backward()
>>> print_grads()
print_grads()
a: 1.0, b: 1.0, c: 1.0, d: 5.0, e: 1
>>> 
>>> e.backward()
e.backward()
>>> print_grads()
print_grads()
a: 3.0, b: 3.0, c: 2.0, d: 10.0, e: 1
>>> 
```

**Comment by eggsyntax** [comment]
This is why I wish everyone would write ML code in a nice sane functional language like Clojure where values aren't getting mutated out from under you all the time :wink:

**Comment by eggsyntax** [comment]
Oh, that's funny, I just got to (nearly) the end of the video, and saw that in fact you'd forgotten to do a zeroing step in the calling code. FWIW to me it seems like it makes sense to include it in the backprop code to ensure callers **can't** forget -- but I'm also new to this; maybe there are good reasons not to incorporate it there.

**Comment by karpathy** [comment]
So actually this is ~intentional and a feature not a bug. This is also how PyTorch works. The reason for this is that in many situations you may have multiple loss functions that you want to backward independently, or you may want to process multiple microbatches before doing the actual update. In PyTorch it is left up to the user to call optimizer.zero_grad() to set the grads to zero for this reason. And that's why I followed it in this repo too. It's just that in this repo specifically we're not seeing the benefits of this design, and the reasons for doing so. I agree it is something one may trip over though, and I've done that multiple times in practice too.

**Comment by eggsyntax** [comment]
Ahh, that makes sense.  Thanks for clarifying! Closing the PR 😄

**Comment by itajaja** [comment]
Still isn't [this code](https://github.com/karpathy/micrograd/blob/c911406e5ace8742e5841a7e0df113ecb5d54685/micrograd/nn.py#L6) misleading? When would I want to JUST zero the grads of the parameters but not of all the internal nodes? Wouldn't I always want to zero everything?

**Comment by ziolko** [comment]
@itajaja Every time Python runs e.g. `act = sum((wi*xi for wi,xi in zip(self.w, x)), self.b)` it creates from scratch intermediate `Value` instances for `*` and `+` operations (what you call "internal nodes"). This is different for parameters (e.g. `wi`) because they defined externally and keep living through whole learning.
